Warn about required environment variables when using LevelZero

cmcantalupo opened the pull request:
- Fixes #3539

avilcheslopez reviewed (dismissed):
Should the documentation (geopm_pio_levelzero(7)) be updated as well? Now that both paths (direct, through service) set it automatically, it seems the user no longer has to make sure it's set.

Also, with this change, does the service still need to set it (in libgeopmd/geopm.service)? If not, it might be a good idea to just set it in one place in case in the future we need to change that setting somehow.

bgeltz reviewed (changes requested):
Re-jig to emit warnings instead of setenv calls.

This violates process expectations about setenv/getenv timing.

cmcantalupo committed:
Clean up warning messages for PlatformTopo creation

- Do not emit a warning if there are no accelerators
- Emit a warning if LevelZero is not set properly

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on libgeopmd/src/GPUTopo.cpp:
```diff
@@ -48,11 +53,16 @@ namespace geopm
-        }
-        catch (const Exception &ex) {
-            // LevelZeroTopo cannot be created, already set to GPUTopoNull
+        if (getenv("ZE_ENABLE_SYSMAN") != nullptr &&
+            std::string(getenv("ZE_ENABLE_SYSMAN")) == "1" &&
+            getenv("ZE_FLAT_DEVICE_HIERARCHY") != nullptr &&
+            std::string(getenv("ZE_FLAT_DEVICE_HIERARCHY")) == "COMPOSITE") {
```
This is a behavior change.  Not 100% sure it is for the better...

cmcantalupo commented on libgeopmd/src/LevelZero.cpp:
```diff
@@ -32,7 +33,30 @@ namespace geopm
+                std::cerr << "LevelZero features require the environment variable setting \"ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE\".\n";
+            }
+            if (do_warn) {
+                std::cerr << "                 This process may corrupt the topology cache file.  Run the following commands as root to fix:\n"
+                          << "                     rm -f /run/geopm/geopm-topo-cache; ZE_ENABLE_SYSMAN=1 ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE geopmread --cache\n";
```
Might be nice to only warn if we are actually creating a topo cache file, but plumbing that turns out to be a bit tricky.

cmcantalupo commented:
```
cmcantal@saruman:~/Git/geopm/libgeopmd (issue-3539)]$ ./geopmread > /dev/null
cmcantal@saruman:~/Git/geopm/libgeopmd (issue-3539)]$ sudo ./geopmread > /dev/null
Warning: <geopm> LevelZero features require the environment variable setting "ZE_ENABLE_SYSMAN=1".
                 LevelZero features require the environment variable setting "ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE".
                 This process may corrupt the topology cache file if LevelZero enabled GPUs are present.
                 Run the following commands to fix:
                     sudo rm -f /run/geopm/geopm-topo-cache
                     sudo env ZE_ENABLE_SYSMAN=1 ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE geopmread --cache

cmcantal@saruman:~/Git/geopm/libgeopmd (issue-3539)]$ sudo env ZE_ENABLE_SYSMAN=1 ./geopmread > /dev/null
Warning: <geopm> LevelZero features require the environment variable setting "ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE".
                 This process may corrupt the topology cache file if LevelZero enabled GPUs are present.
                 Run the following commands to fix:
                     sudo rm -f /run/geopm/geopm-topo-cache
                     sudo env ZE_ENABLE_SYSMAN=1 ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE geopmread --cache
cmcantal@saruman:~/Git/geopm/libgeopmd (issue-3539)]$ sudo env ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE  ./geopmread > /dev/null
Warning: <geopm> LevelZero features require the environment variable setting "ZE_ENABLE_SYSMAN=1".
                 This process may corrupt the topology cache file if LevelZero enabled GPUs are present.
                 Run the following commands to fix:
                     sudo rm -f /run/geopm/geopm-topo-cache
                     sudo env ZE_ENABLE_SYSMAN=1 ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE geopmread --cache
```

bgeltz reviewed (commented):
Changes seem fine.  This should be tested on a L0 enabled system w/appropriate hardware, including building and deploying the service RPMs before we merge.

cmcantalupo committed:
Fix typo

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
```
$ geopmread GPU_POWER gpu 0
286.8566902140415
$ sudo geopmread
Warning: <geopm> LevelZero features require the environment variable setting "ZES_ENABLE_SYSMAN=1".
                 LevelZero features require the environment variable setting "ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE".
                 This process may corrupt the topology cache file if LevelZero enabled GPUs are present.
                 Run the following commands to fix:
                     sudo rm -f /run/geopm/geopm-topo-cache
                     sudo env ZES_ENABLE_SYSMAN=1 ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE geopmread --cache
...
```

also a clean syslog on that run.
